### PR TITLE
feat: created multimeter activity

### LINKED ISSUE #885: Create the multimeter module
**Actual Behaviour**

Currently the multimeter is inside the control unit section.

**Expected Behaviour**

It has to be made a separate module itself.

**Steps to reproduce it**

Create a multimeter fragment inside the instrument fragment and implement all it's functionality.

**LogCat for the issue**

Not applicable

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by abhinavraj23**
Fixes #885 

Changes: Created the multimeter activity

Screenshots for the change: 
multimeter

APK for testing: [Compress the *.apk file into a rar or zip file and upload it here]:The apk will only make sense if the pr #884 is merged as there is not different multimeter section in the previous version of app so there is no button to open the activity.

### CODE REVIEW

**Comment by abhinavraj23** [comment]
@CloudyPadmal Sorry for closing the previous pr, there was some issue with it. In this PR I have updated the layout, please review the layout and let me know if there are any changes to be done in it Please note that there are still many changes to be done in the code part(eg removing hard-coded dimensions) so please ignore that for the moment and let me know if I can continue with this layout..

**Comment by harsh-2711** [comment]
@abhinavraj23  What's that blue box on text "Voltage of Channels" ?

**Comment by abhinavraj23** [comment]
@harsh-2711 Thanks for the review about the layout, I will although update it(as it has hard-coded dimensions).

**Comment by CloudyPadmal** [comment]
@abhinavraj23 you could have updated the changes in the same PR without closing and creating new PRs. I see you haven't even made the changes suggested in the closed PR in this PR. Please go through the closed PR to see the suggested changes.

**Comment by abhinavraj23** [comment]
@CloudyPadmal I would do all the changes but first I wanted to make sure whether or not this layout is good! If you are fine with this layout then I would make all the necessary changes :+1:

**Comment by Vikum94** [comment]
Layout seems ok :+1:

**Comment by abhinavraj23** [comment]
@CloudyPadmal I have extracted all the hard-coded dimensions, I will update the java file by tomorrow :+1:

**Comment by CloudyPadmal** [comment]
That's great! :+1:

**Comment by abhinavraj23** [comment]
@CloudyPadmal I have implemented the java codes thus the multimeter activity is now completed. But till now we don't have any button to start the activity so I am waiting for pr #884 to be merged.Also I would like you to note that the extra changes came while solving the merge conflict

**Comment by CloudyPadmal** [comment]
@abhinavraj23 #884 is now merged!

File: app/src/main/res/layout/activity_multimeter.xml [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+            android:id="@+id/quantity"
+            android:layout_width="wrap_content"
+            android:layout_height="@dimen/multimeter_display_text_margin"
+            android:layout_alignParentRight="true"
+            android:layout_marginRight="@dimen/multimeter_layout_margin"
+            android:layout_marginTop="@dimen/multimeter_display_text_margin"
+            android:background="@android:color/white"
+            android:fontFamily="sans-serif"
+            android:gravity="center"
+            android:text="200"
```
Comment (Vikum94) [inline app/src/main/res/layout/activity_multimeter.xml]: Is this a test line. This is not needed is it?

File: app/src/main/res/layout/activity_multimeter.xml [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+        <TextView
+            android:id="@+id/unit"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_centerHorizontal="true"
+            android:layout_marginTop="@dimen/multimeter_unitbox_topmargin"
+            android:background="@android:color/white"
+            android:fontFamily="sans-serif"
+            android:gravity="center"
+            android:text="Volts"
```
Comment (Vikum94) [inline app/src/main/res/layout/activity_multimeter.xml]: This goes in the string resource file.

File: app/src/main/res/layout/activity_multimeter.xml (hunk 1 above)
Comment (abhinavraj23) [reply app/src/main/res/layout/activity_multimeter.xml]: Yes, it is a test line, will remove it :+1:

File: app/src/main/res/layout/activity_multimeter.xml (hunk 2 above)
Comment (abhinavraj23) [reply app/src/main/res/layout/activity_multimeter.xml]: This will also be removed

**Comment by CloudyPadmal** [comment]
@abhinavraj23 are you having troubles resolving conflicts?

**Comment by CloudyPadmal** [comment]
@abhinavraj23 try fetching the upstream and from your `multimeter` branch, run 
`$ git reset --soft upstream/development`

It will clear all the commits related to `multimeter` branch and get the HEAD back so you can force push one commit. Let me know the progress.

**Comment by abhinavraj23** [comment]
@CloudyPadmal Sure, but i I do that will it not kill all my changes made?

File: app/build.gradle [hunk 3]
```diff
@@ -44,6 +44,9 @@ dependencies {
     implementation "com.github.devlight.navigationtabstrip:navigationtabstrip:$rootProject.navTabStripVersion"
     implementation "com.afollestad.material-dialogs:core:$rootProject.materialDialogsVersion"
     implementation "com.jakewharton:butterknife:$rootProject.butterKnifeVersion"
+
+    implementation 'com.github.BeppiMenozzi:Knob:1.9.0'
```
Comment (CloudyPadmal) [inline app/build.gradle]: OK @abhinavraj23 now comes reviews! :D Add `1.9.0` to the gradle config file just like the other libraries.

**Comment by abhinavraj23** [comment]
Thanks a lot @CloudyPadmal . It really helped me :+1:

File: app/build.gradle [hunk 4]
```diff
@@ -44,6 +44,9 @@ dependencies {
     implementation "com.github.devlight.navigationtabstrip:navigationtabstrip:$rootProject.navTabStripVersion"
     implementation "com.afollestad.material-dialogs:core:$rootProject.materialDialogsVersion"
     implementation "com.jakewharton:butterknife:$rootProject.butterKnifeVersion"
+
```
Comment (CloudyPadmal) [inline app/build.gradle]: And remove these extra spaces

File: app/src/main/java/org/fossasia/pslab/activity/MultimeterActivity.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    private int knobState;
+
+    private String[] knobMarker = {"ID1","ID2","ID3","ID4","ID4","ID3","ID2","ID1","CH3","CH2","CH1","CAP","SEN","AN8"};
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_multimeter);
+        scienceLab = ScienceLabCommon.scienceLab;
+        Button readResistance = (Button) findViewById(R.id.resistance);
```
Comment (CloudyPadmal) [inline app/src/main/java/org/fossasia/pslab/activity/MultimeterActivity.java]: Use butterknife for view binding

**Comment by CloudyPadmal** [comment]
Also go through Codacy review and try to fix those issues as well

**Comment by abhinavraj23** [comment]
@CloudyPadmal I have changed the linear layout to relative layout as initially the layout was varying in different mobiles and the finally its done.Also I suppose now the codacy will be working Here is the apk.
[app-debug.apk.zip](https://github.com/fossasia/pslab-android/files/2048208/app-debug.apk.zip)

**Comment by abhinavraj23** [comment]
@CloudyPadmal Now I have also added bind-view so all changes are done. Please review now :+1:

File: app/src/main/java/org/fossasia/pslab/activity/MultimeterActivity.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    @BindView(R.id.read)
+    Button read;
+    @BindView(R.id.capacitance)
+    Button readCapacitance;
+    @BindView(R.id.resistance)
+    Button readResistance;
+
+    private int knobState;
+
+    private String[] knobMarker = {"ID1", "ID2", "ID3", "ID4", "ID4", "ID3", "ID2", "ID1", "CH3", "CH2", "CH1", "CAP", "SEN", "AN8"};
```
Comment (harsh-2711) [inline app/src/main/java/org/fossasia/pslab/activity/MultimeterActivity.java]: Use resources file for adding string array

File: app/src/main/java/org/fossasia/pslab/activity/MultimeterActivity.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import it.beppi.knoblibrary.Knob;
+
+/**
+ * Created by Abhinav Raj on 26/5/18.
+ */
+
+
```
Comment (harsh-2711) [inline app/src/main/java/org/fossasia/pslab/activity/MultimeterActivity.java]: No need of extra lines

**Comment by abhinavraj23** [comment]
@harsh-2711 Changes done :+1:

**Comment by abhinavraj23** [comment]
@CloudyPadmal Please remove the WIP tag as the work is complete now, please add the review-required tag.

**Comment by CloudyPadmal** [comment]
@abhinavraj23 this is how UI looks in my device.
screen

**Comment by abhinavraj23** [comment]
Okay I l solve it right know +1

**Comment by abhinavraj23** [comment]
@CloudyPadmal @harsh-2711 @Vikum94 @Avjeet I have made the changes and have tried to make a common layout but please test it in your devices if a problem occurs then do tell me.

File: app/src/main/java/org/fossasia/pslab/fragment/HomeFragment.java [hunk 8]
```diff
@@ -80,56 +64,21 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
-            }
-            imgViewDeviceStatus.setImageResource(R.drawable.icons8_usb_connected_100);
+            imgViewDeviceStatus.setImageResource(org.fossasia.pslab.R.drawable.usb_connected);
             tvDeviceStatus.setText(getString(R.string.device_connected_successfully));
         } else {
-            imgViewDeviceStatus.setImageResource(R.drawable.icons_usb_disconnected_100);
+            imgViewDeviceStatus.setImageResource(org.fossasia.pslab.R.drawable.usb_disconnected);
             tvDeviceStatus.setText(getString(R.string.device_not_found));
         }
-        deviceDescription.setOnClickListener(new View.OnClickListener() {
```
Comment (Avjeet) [inline app/src/main/java/org/fossasia/pslab/fragment/HomeFragment.java]: This file was added in PR https://github.com/fossasia/pslab-android/pull/897

**Comment by abhinavraj23** [comment]
Changes done @Avjeet

**Comment by Avjeet** [comment]
@abhinavraj23 could u provide with the apk .

File: app/src/main/res/values/dimens.xml [hunk 9]
```diff
@@ -89,11 +89,11 @@
     <dimen name="seekbar_height">16dp</dimen>
 
     <dimen name="home_fragment_padding">10dp</dimen>
-    <dimen name="home_fragment_margin">30dp</dimen>
```
Comment (Avjeet) [inline app/src/main/res/values/dimens.xml]: plz remove this change also, this is also from the previous PR #897

**Comment by abhinavraj23** [comment]
[Uploading app-debug.apk.zip…]() It is the latest apk

**Comment by harsh-2711** [comment]
@abhinavraj23 The layout doesn't looks good on big screen mobiles so try to use linear layout as parent layout so that no extra spaces are left. 
screenshot_20180602-033316

Also the COUNT PULSE label is only in capitals whereas others are in normal case. Please fix this. Also try to give some space between knob and text. If possible try to use transparent buttons for the text around the knob so that the knob can be rotated on clicking the text. Keep all the main headings i.e. Frequency, Voltage at channels.. center aligned and change their style to bold and increase there text size as they should look different from main labels.

**Comment by CloudyPadmal** [comment]
@abhinavraj23 please focus on this instrument first and complete it. I will take up Power Sources. We are already at the end of [Milestone - 1](https://github.com/fossasia/pslab-android/milestone/17)

**Comment by abhinavraj23** [comment]
@harsh-2711 @CloudyPadmal now it works on all the mobiles and it is perfect. Here is the latest apk
[app-debug.apk.zip](https://github.com/fossasia/pslab-android/files/2064979/app-debug.apk.zip)

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -668,5 +685,15 @@
 
     <string name="text_connected">Connected</string>
     <string name="text_disconnected">Disconnected</string>
+    <string name="voltage_channels">Voltage at Channels</string>
```
Comment (CloudyPadmal) [inline app/src/main/res/values/strings.xml]: @abhinavraj23 can you add these strings somewhere in the middle, may be next to `multimeter_knob_states`? Adding content to resource files at the end makes a lot of merge conflicts in almost every PR.

File: app/src/main/res/values/dimens.xml [hunk 11]
```diff
@@ -133,4 +164,5 @@
     <dimen name="progress_bar_height">8dp</dimen>
     <dimen name="connect_instruction_text_header">18sp</dimen>
 
-</resources>
```
Comment (CloudyPadmal) [inline app/src/main/res/values/dimens.xml]: Also revert changes done here at the end of the file. It's safe not to do any modifications to the end of resource files.

**Comment by CloudyPadmal** [comment]
OK APK looks good. @abhinavraj23 can you have the labels placed in one order. Here we have CH1, CH2, CH3 and then ID1, ... , ID4 and then again starting from ID4 ... Please make it ID1 -> ID4 in the ascending order.

**Comment by abhinavraj23** [comment]
@CloudyPadmal Please test this new apk I have made some minor changes
[app-debug.apk.zip](https://github.com/fossasia/pslab-android/files/2065000/app-debug.apk.zip)

@CloudyPadmal do you need order like ID1,ID2....ID4,ID1,ID2...ID4 ?
I think the current order looks good due to its symmetry.

**Comment by CloudyPadmal** [comment]
Yes please change the order to ID1,ID2....ID4,ID1,ID2...ID4. I'll check the APK

**Comment by abhinavraj23** [comment]
@CloudyPadmal All changes done :+1:

**Comment by CloudyPadmal** [comment]
@abhinavraj23 Now it's all good. Please address the changes requested and let's complete this PR! :+1:

**Comment by abhinavraj23** [comment]
@CloudyPadmal All changes done :+1:

**Comment by abhinavraj23** [comment]
@Avjeet I have changed the read button to the right side but I think colour of the display screen is okay and is according to the theme. @harsh-2711 @Vikum94 Please review as this PR is taking lot of time.

**Pull request merged**